mypyc/irbuild/classdef.py

```diff
@@ -108,53 +105,207 @@ def transform_class_def(builder: IRBuilder, cdef: ClassDef) -> None:
                 continue
             # We want to collect class variables in a dictionary for both real
             # non-extension classes and fake dataclass ones.
-            var_non_ext = non_ext or dataclass_non_ext
-            if var_non_ext:
-                add_non_ext_class_attr(builder, var_non_ext, lvalue, stmt, cdef, attrs_to_cache)
-                if non_ext:
-                    continue
-            # Variable declaration with no body
-            if isinstance(stmt.rvalue, TempNode):
-                continue
-            # Only treat marked class variables as class variables.
-            if not (is_class_var(lvalue) or stmt.is_final_def):
-                continue
-            typ = builder.load_native_type_object(cdef.fullname)
-            value = builder.accept(stmt.rvalue)
-            builder.call_c(
-                py_setattr_op, [typ, builder.load_str(lvalue.name), value], stmt.line)
-            if builder.non_function_scope() and stmt.is_final_def:
-                builder.init_final_static(lvalue, value, cdef.name)
+            cls_builder.add_attr(lvalue, stmt)
+
         elif isinstance(stmt, ExpressionStmt) and isinstance(stmt.expr, StrExpr):
             # Docstring. Ignore
             pass
         else:
             builder.error("Unsupported statement in class body", stmt.line)
 
-    if not non_ext:  # That is, an extension class
-        generate_attr_defaults(builder, cdef)
-        create_ne_from_eq(builder, cdef)
-        if dataclass_non_ext:
-            assert type_obj
-            dataclass_finalize(builder, cdef, dataclass_non_ext, type_obj)
-    else:
+    cls_builder.finalize(ir)
+
+
+class ClassBuilder:
+    """Create IR for a class definition."""
```

Comment:
Thanks for refactoring these! This now looks cleaner.

Can you mention that this is effectively an abstract class in the docstring, even though there are no abstract members?

Reply:
done.  I also went ahead and marked the methods as abstract:  even though they all return None a subclass would effectively be broken if it did not implement these (with the possible exception of `finalize`)